pkg/apigee/apigee.go

```diff
@@ -189,13 +191,165 @@ func (c *Client) hostAliases(org, env, virtualHost string) ([]string, error) {
 	return hosts, nil
 }
 
+// CPSEnabledForOrg returns a bool indicating whether the requested Org has CPS enabled or not
+func (c *Client) CPSEnabledForOrg(orgName string) (bool, error) {
+	c.initDefaults()
+
+	cpsEnabled := false
+
+	orgURL := fmt.Sprintf("%sv1/organizations/%s", c.ApigeeAPIHost, orgName)
+	resp, err := c.Get(orgURL)
+	if err != nil {
+		return false, err
+	}
+
+	defer resp.Body.Close()
+
+	var rawOrg interface{}
+
+	err = json.NewDecoder(resp.Body).Decode(&rawOrg)
+
+	if err != nil {
+		fmt.Printf("Error unmarshalling response: %v\n", err)
+		return false, err
+	}
+
+	org := rawOrg.(map[string]interface{})
+	orgProps := org["properties"].(map[string]interface{})
+	orgProp := orgProps["property"].([]interface{})
+
+	for _, rawProp := range orgProp {
+		prop := rawProp.(map[string]interface{})
+		if prop["name"] == "features.isCpsEnabled" {
+			if prop["value"] == "true" {
+				return true, nil
+			}
+			break
+		}
+	}
+	return cpsEnabled, nil
+}
+
+func (c *Client) CreateKVM(orgName, envName, publicKey string) error {
+	const apigeeKVMName = "shipyard-routing"
+	const apigeeKVMPKName = "x-routing-api-key"
+
+
+	httpClient := &http.Client{}
+
+	//construct URL
+	kvmURL := fmt.Sprintf("%sv1/organizations/%s/environments/%s/keyvaluemaps", c.ApigeeAPIHost, orgName, envName)
+
+	//create JSON body
+	kvmBody := apigeeKVMBody{
+		Name: apigeeKVMName,
+		Entry: []apigeeKVMEntry{
+			{
+				Name:  apigeeKVMPKName,
+				Value: base64.StdEncoding.EncodeToString([]byte(publicKey)),
+			},
+		},
+	}
+
+	b := new(bytes.Buffer)
+	json.NewEncoder(b).Encode(kvmBody)
+
+	req, err := http.NewRequest("POST", kvmURL, b)
+	if err != nil {
+		errorMessage := fmt.Sprintf("Unable to create request (Create KVM): %v", err)
+		return errors.New(errorMessage)
+	}
+
+	//Must pass through the authz header
+	req.Header.Add("Authorization", c.Token)
+	req.Header.Add("Content-Type", "application/json")
+
+	resp, err := httpClient.Do(req)
+	if err != nil {
+		errorMessage := fmt.Sprintf("Error creating Apigee KVM: %v", err)
+		return errors.New(errorMessage)
+	}
+	defer resp.Body.Close()
+
+	// If the response was not a 201, we need to check if the response was a 409 because this means the KVM exists
+	// already and we'll need to update the KVM value(s).
+	if resp.StatusCode != 201 {
+		var retryFlag bool
+
+		// If the KVM already exists, we need to update its value(s).
+		if resp.StatusCode == 409 {
+			b2 := new(bytes.Buffer)
+			updateKVMURL := fmt.Sprintf("%s/%s", kvmURL, apigeeKVMName) // Use non-CPS endpoint by default
+
+			cpsBool, err := c.CPSEnabledForOrg(orgName)
```

Comment:
Why not `cpsEnabled`?  Putting the type into the variable name doesn't seem ideal.